src/main/tut/3.5-monads-and-effects.md

```diff
@@ -0,0 +1,446 @@
+    val sb = k(a)
+    runState(sb)(s1)
+  }
+```
+................
+```Haskell
+instance Monad (State s) where
+    sa >>= k = State (\s -> let (a, s') = runState sa s
+                            in runState (k a) s')
+    return a = State (\s -> (a, s))
+```
+```tut:silent
+implicit def stateMonad[S] =
+  new Monad[({type T[A] = State[S, A]})#T] {
+    def flatMap[A, B]
+        (sa: State[S, A])
+        (k: A => State[S, B])
+        : State[S, B] =
+      State { s =>
+        val (a, s1) = runState(sa)(s)
+        runState(k(a))(s1)
+      }
+
+    def pure[A](a: A): State[S, A] =
+      State(s => (a, s))
+  }
+```
+................
+```Haskell
+get :: State s s
+get = State (\s -> (s, s))
+```
+```scala
+def get: State[S, S] =
+  State(s => (s, s))
+```
+................
+```Haskell
+put :: s -> State s ()
+put s' = State (\s -> ((), s'))
+```
+```scala
+def put(s1: S): State[S, Unit] =
+  State(s => ((), s1))
+```
+................
+```Haskell
+instance Monad Maybe where
+    Nothing >>= k = Nothing
+    Just a  >>= k = k a
+    return a = Just a
+```
+```tut:silent
+implicit val optionMonad =
+  new Monad[Option] {
+    def flatMap[A, B]
+        (ma: Option[A])(k: A => Option[B])
+        : Option[B] = ma match {
+      case None => None
+      case Some(a) => k(a)
+    }
+
+    def pure[A](a: A): Option[A] = Some(a)
+  }
+```
+................
+```Haskell
+data Cont r a = Cont ((a -> r) -> r)
+```
+```tut:silent
+case class Cont[R, A](run: (A => R) => R)
+```
+................
+```Haskell
+runCont :: Cont r a -> (a -> r) -> r
+runCont (Cont k) h = k h
+```
+```tut:silent
+def runCont[R, A]: Cont[R, A] => (A => R) => R = {
+  case Cont(k) => h => k(h)
+}
+```
+................
+```Haskell
+(>>=) :: ((a -> r) -> r) ->
+         (a -> (b -> r) -> r) ->
+         ((b -> r) -> r)
+```
+```scala
+def flatMap[A, B]
+    : ((A => R) => R) =>
+      (A => (B => R) => R) =>
+      ((B => R) => R)
+```
+................
+```Haskell
+ka >>= kab = Cont (\hb -> ...)
+```
+```scala
+def flatMap[A, B](ka: Cont[R, A])(kab: A => Cont[R, B]): Cont[R, B] =
+  Cont(hb => ???)
+```
+................
+```Haskell
+runCont ka (\a -> ...)
+```
+```scala
+runCont(ka)(a => ???)
+```
+................
+```Haskell
+runCont ka (\a -> let kb = kab a
+                  in runCont kb hb)
+```
+```scala
+runCont(ka) { a =>
+  val kb = kab(a)
+  runCont(kb)(hb)
+}
+```
+................
+```Haskell
+instance Monad (Cont r) where
+    ka >>= kab = Cont (\hb -> runCont ka (\a -> runCont (kab a) hb))
+    return a = Cont (\ha -> ha a)
+```
+```tut:silent
+implicit def contMonad[R] =
+  new Monad[({type T[A] = Cont[R, A]})#T] {
+    def flatMap[A, B]
+        (ka: Cont[R, A])
+        (kab: A => Cont[R, B])
+        : Cont[R, B] =
+      Cont { hb =>
+        runCont(ka)(a => runCont(kab(a))(hb))
+      }
+
+    def pure[A](a: A): Cont[R, A] =
+      Cont(ha => ha(a))
+  }
+```
+................
+```Haskell
+main :: IO ()
+```
+```scala
+def main: IO[Unit]
```

Comment:
does this part need some comments? that Scala is not Haskell, or is it clear?

Reply:
looks fine to me.